Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift

```diff
@@ -735,6 +742,40 @@ extension SwiftLanguageServer {
     }
     ranges.append(range)
   }
+
+  func codeAction(_ req: Request<CodeActionRequest>) {
+    let providersAndKinds: [(provider: CodeActionProvider, kind: CodeActionKind)] = [
+      //TODO: Implement the providers.
+      //(retrieveRefactorCodeActions, .refactor),
+      //(retrieveQuickFixCodeActions, .quickFix)
+    ]
+    let wantedActionKinds = req.params.context.only
+    let providers = providersAndKinds.filter { wantedActionKinds?.contains($0.1) != false }
+    retrieveCodeActions(req, providers: providers.map { $0.provider }) { codeActions in
+      let capabilities = self.clientCapabilities.textDocument?.codeAction
+      let response = CodeActionRequestResponse(codeActions: codeActions,
+                                               clientCapabilities: capabilities)
+      req.reply(response)
+    }
+  }
+
+  func retrieveCodeActions(_ req: Request<CodeActionRequest>, providers: [CodeActionProvider], completion: @escaping CodeActionProviderCompletion) {
+    guard providers.isEmpty == false else {
+      completion([])
+      return
+    }
+    var providersLeftToProcess = Set<Int>(0..<providers.count)
+    var codeActions = [CodeAction]()
+    for i in 0..<providers.count {
+      providers[i](req.params) { actions in
+        codeActions += actions
+        providersLeftToProcess.remove(i)
```

Comment:
Why do we need `providersLeftToProcess`?  When we're at iteration `i`, either we have `providersLeftToProcess == i...providers.count`, or we have somehow missed one and we will exit without calling the completion handler, which is a logic error.

Reply:
The ones that call sourcekitd can be async, so we keep track of who's left because we don't know the order they're going to finish -- is that what you meant or I misunderstood the comment? If the former, I think I'll add an operation queue to make it less hacky.

Reply:
I see. What happens if there's an error?

Reply:
I'm having them log the error and return an empty list to not affect the other code actions:
https://github.com/apple/sourcekit-lsp/blob/f0b632c6081740d6ead21a398a56b26103f852b4/Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift#L698

Reply:
Okay, the critical thing is that we always call the request's reply method eventually.

Reply:
Sure, we can make a test to prove it when a provider is added. I had to make the completion optional for now to avoid a compilation issue, but once we add the refactoring provider I can iron this out.